File: easy-random-bean-validation/src/test/java/org/jeasy/random/validation/PastAnnotationHandlerTest.java
package org.jeasy.random.validation;

import static org.assertj.core.api.Assertions.assertThat;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.Past;
import java.time.*;
import java.time.chrono.HijrahDate;
import java.time.chrono.JapaneseDate;
import java.time.chrono.MinguoDate;
import java.time.chrono.ThaiBuddhistDate;
import java.util.Calendar;
import java.util.Date;
import java.util.Set;
import org.jeasy.random.EasyRandom;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PastAnnotationHandlerTest {

  private EasyRandom easyRandom;

  @BeforeEach
  void setUp() {
    easyRandom = new EasyRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    // given
    // when
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    // then
    assertThat(testBean.getTestDate()).isBefore(Date.from(Instant.now()));
    assertThat(testBean.getTestCalendar()).isLessThan(Calendar.getInstance());
    assertThat(testBean.getTestInstant()).isBefore(Instant.now());
    assertThat(testBean.getTestLocalDate()).isBefore(LocalDate.now());
    assertThat(testBean.getTestLocalDateTime()).isBefore(LocalDateTime.now());
    assertThat(testBean.getTestLocalTime()).isBefore(LocalTime.now());
    assertThat(testBean.getTestMonthDay()).isLessThan(MonthDay.now());
    assertThat(testBean.getTestOffsetDateTime()).isBefore(OffsetDateTime.now());
    assertThat(testBean.getTestOffsetTime()).isBefore(OffsetTime.now());
    assertThat(testBean.getTestYear()).isLessThan(Year.now());
    assertThat(testBean.getTestYearMonth()).isLessThan(YearMonth.now());
    assertThat(testBean.getTestZonedDateTime()).isBefore(ZonedDateTime.now());
    assertThat(testBean.getTestHijrahDate()).isLessThan(HijrahDate.now());
    assertThat(testBean.getTestJapaneseDate()).isLessThan(JapaneseDate.now());
    assertThat(testBean.getTestMinguoDate()).isLessThan(MinguoDate.now());
    assertThat(testBean.getTestThaiBuddhistDate()).isLessThan(ThaiBuddhistDate.now());
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    // given
    // when
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    // then
    assertThat(violations).isEmpty();
  }

  static class TestBean {
    @Past private Date testDate;

    public Date getTestDate() {
      return testDate;
    }

    @Past private Calendar testCalendar;

    public Calendar getTestCalendar() {
      return testCalendar;
    }

    @Past private Instant testInstant;

    public Instant getTestInstant() {
      return testInstant;
    }

    @Past private LocalDate testLocalDate;

    public LocalDate getTestLocalDate() {
      return testLocalDate;
    }

    @Past private LocalDateTime testLocalDateTime;

    public LocalDateTime getTestLocalDateTime() {
      return testLocalDateTime;
    }

    @Past private LocalTime testLocalTime;

    public LocalTime getTestLocalTime() {
      return testLocalTime;
    }

    @Past private MonthDay testMonthDay;

    public MonthDay getTestMonthDay() {
      return testMonthDay;
    }

    @Past private OffsetDateTime testOffsetDateTime;

    public OffsetDateTime getTestOffsetDateTime() {
      return testOffsetDateTime;
    }

    @Past private OffsetTime testOffsetTime;

    public OffsetTime getTestOffsetTime() {
      return testOffsetTime;
    }

    @Past private Year testYear;

    public Year getTestYear() {
      return testYear;
    }

    @Past private YearMonth testYearMonth;

    public YearMonth getTestYearMonth() {
      return testYearMonth;
    }

    @Past private ZonedDateTime testZonedDateTime;

    public ZonedDateTime getTestZonedDateTime() {
      return testZonedDateTime;
    }

    @Past private HijrahDate testHijrahDate;

    public HijrahDate getTestHijrahDate() {
      return testHijrahDate;
    }

    @Past private JapaneseDate testJapaneseDate;

    public JapaneseDate getTestJapaneseDate() {
      return testJapaneseDate;
    }

    @Past private MinguoDate testMinguoDate;

    public MinguoDate getTestMinguoDate() {
      return testMinguoDate;
    }

    @Past private ThaiBuddhistDate testThaiBuddhistDate;

    public ThaiBuddhistDate getTestThaiBuddhistDate() {
      return testThaiBuddhistDate;
    }
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/PastAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.comparables.shouldBeLessThan
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.Past
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.MonthDay
import java.time.OffsetDateTime
import java.time.OffsetTime
import java.time.Year
import java.time.YearMonth
import java.time.ZonedDateTime
import java.time.chrono.HijrahDate
import java.time.chrono.JapaneseDate
import java.time.chrono.MinguoDate
import java.time.chrono.ThaiBuddhistDate
import java.util.Calendar
import java.util.Date
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class PastAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun `generated bean should be valid according to validation constraints`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.testDate shouldBeLessThan Date.from(Instant.now())
    testBean.testCalendar shouldBeLessThan Calendar.getInstance()
    testBean.testInstant shouldBeLessThan Instant.now()
    testBean.testLocalDate shouldBeLessThan LocalDate.now()
    testBean.testLocalDateTime shouldBeLessThan LocalDateTime.now()
    testBean.testLocalTime shouldBeLessThan LocalTime.now()
    testBean.testMonthDay shouldBeLessThan MonthDay.now()
    testBean.testOffsetDateTime shouldBeLessThan OffsetDateTime.now()
    testBean.testOffsetTime shouldBeLessThan OffsetTime.now()
    testBean.testYear shouldBeLessThan Year.now()
    testBean.testYearMonth shouldBeLessThan YearMonth.now()
    testBean.testZonedDateTime shouldBeLessThan ZonedDateTime.now()
    testBean.testHijrahDate shouldBeLessThan HijrahDate.now()
    testBean.testJapaneseDate shouldBeLessThan JapaneseDate.now()
    testBean.testMinguoDate shouldBeLessThan MinguoDate.now()
    testBean.testThaiBuddhistDate shouldBeLessThan ThaiBuddhistDate.now()
  }

  @Test
  fun `generated bean should be valid using bean validation api`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(
    @field:Past val testDate: Date,
    @field:Past val testCalendar: Calendar,
    @field:Past val testInstant: Instant,
    @field:Past val testLocalDate: LocalDate,
    @field:Past val testLocalDateTime: LocalDateTime,
    @field:Past val testLocalTime: LocalTime,
    @field:Past val testMonthDay: MonthDay,
    @field:Past val testOffsetDateTime: OffsetDateTime,
    @field:Past val testOffsetTime: OffsetTime,
    @field:Past val testYear: Year,
    @field:Past val testYearMonth: YearMonth,
    @field:Past val testZonedDateTime: ZonedDateTime,
    @field:Past val testHijrahDate: HijrahDate,
    @field:Past val testJapaneseDate: JapaneseDate,
    @field:Past val testMinguoDate: MinguoDate,
    @field:Past val testThaiBuddhistDate: ThaiBuddhistDate,
  )
}
